module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1987 @@
+ * AES-CCM. Its purpose is to ensure that an attacker cannot modify encrypted
+ * data on disk and return garbage to the application. Effectively, it is a
+ * checksum that can not be reproduced by an attacker. We store the MAC in the
+ * second 128 bits of blk_cksum, leaving the first 128 bits for a truncated
+ * regular checksum of the ciphertext which can be used for scrubbing.
+ *
+ * OBJECT AUTHENTICATION:
+ * Some object types, such as DMU_OT_MASTER_NODE cannot be encrypted because
+ * they contain some info that always needs to be readable. To prevent this
+ * data from being altered, we authenticate this data using SHA512-HMAC. This
+ * will produce a MAC (similar to the one produced via encryption) which can
+ * be used to verify the object was not modified. HMACs do not require key
+ * rotation or IVs, so we can keep all 3 copies of authenticated data.
+ *
+ * ZIL ENCRYPTION:
+ * ZIL blocks have their bp written to disk ahead of the associated data, so we
+ * cannot store encryption paramaters there as we normally do. For these blocks
+ * the MAC is stored in the embedded checksum within the zil_chain_t header. The
+ * salt and IV are generated for the block on bp allocation instead of at
+ * encryption time. In addition, ZIL blocks have some pieces that must be left
+ * in plaintext for claiming while all of the sensitive user data still needs to
+ * be encrypted. The function zio_crypt_init_uios_zil() handles parsing which
+ * which pieces of the block need to be encrypted. All data that is not
+ * encrypted is authenticated using the AAD mechanisms that the supported
+ * encryption modes provide for. In order to preserve the semantics of the ZIL
+ * for encrypted datasets, the ZIL is not protected at the objset level as
+ * described below.
+ *
+ * DNODE ENCRYPTION:
+ * Similarly to ZIL blocks, the core part of each dnode_phys_t needs to be left
+ * in plaintext for scrubbing and claiming, but the bonus buffers might contain
+ * sensitive user data. The function zio_crypt_init_uios_dnode() handles parsing
+ * which which pieces of the block need to be encrypted. For more details about
+ * dnode authentication and encryption, see zio_crypt_init_uios_dnode().
+ *
+ * OBJECT SET AUTHENTICATION:
+ * Up to this point, everything we have encrypted and authenticated has been
+ * at level 0 (or -2 for the ZIL). If we did not do any further work the
+ * on-disk format would be suscptible to attacks that deleted or rearrannged
+ * the order of level 0 blocks. Ideally, the cleanest solution would be to
+ * maintain a tree of authentication MACs going up the bp tree. However, this
+ * presents a problem for raw sends. Send files do not send information about
+ * indirect blocks so there would be no convenient way to transfer the MACs and
+ * they cannot be recalculated on the receive side without the master key which
+ * would defeat the purpose of raw sends in the first place. Instead, for the
```

Comment:
`the purpose` -> `one of the purposes` (having it encrypted over the wire at no CPU cost is still valuable)